Mathlib/NumberTheory/SiegelsLemma.lean

```diff
@@ -0,0 +1,237 @@
+
+private lemma card_S_eq : (Finset.Icc N P).card = (∏ i : Fin m, (P i - N i + 1)):= by
+  rw [Pi.card_Icc N P, Nat.cast_prod]
+  congr
+  ext i
+  rw [Int.card_Icc_of_le (N i) (P i) (N_le_P_add_one m n A i)]
+  exact add_sub_right_comm (P i) 1 (N i)
+
+/-- The sup norm of a non-zero integer matrix is at lest one  -/
+lemma one_le_norm_A_of_ne_zero : 1 ≤ ‖A‖ := by
+  by_contra h
+  apply lt_of_not_le at h
+  apply hA
+  ext i j
+  simp only [zero_apply]
+  rw [norm_lt_iff Real.zero_lt_one] at h
+  specialize h i j
+  rw [Int.norm_eq_abs] at h
+  norm_cast at h
+  apply (Int.abs_lt_one_iff.1 h)
+
+-- # Step 2: S.card < T.card
+
+open Real Nat
+
+private lemma card_S_lt_card_T : (S).card < (T).card := by
+  zify --this is necessary to use card_S_eq
+  rw [card_T_eq, card_S_eq]
+  rify --this is necessary because ‖A‖ is a real number
+  calc
+  ∏ x : Fin m, (∑ x_1 : Fin n, ↑B * ↑(A x x_1)⁺ - ∑ x_1 : Fin n, ↑B * -↑(A x x_1)⁻ + 1)
+    ≤ (n * ‖A‖ * B + 1) ^ m := by
+      rw [← Fin.prod_const m (n * ‖A‖ * B + 1)] --transform ^m into a product
+      apply Finset.prod_le_prod --two goals
+      all_goals intro i _
+      · obtain h:= N_le_P_add_one m n A i
+        rify at h
+        linarith only [h]
+      · simp only [mul_neg, sum_neg_distrib, sub_neg_eq_add, add_le_add_iff_right]
+        have h1 : n * ‖A‖ * B = ∑ _ : Fin n, ‖A‖ * B := by
+          simp only [sum_const, card_univ, Fintype.card_fin, nsmul_eq_mul, (mul_assoc ↑n ‖A‖ ↑B)]
+        rw [h1, ← Finset.sum_add_distrib]
+        gcongr with j _
+        rw [← mul_add, ← Int.cast_add, posPart_add_negPart (A i j), mul_comm]
+        apply mul_le_mul_of_nonneg_right _ (Nat.cast_nonneg B)
+        rw [Int.cast_abs]
+        exact norm_entry_le_entrywise_sup_norm A
+  _  ≤ (n * ‖A‖) ^ m * (B + 1) ^ m := by
+        rw [← mul_pow]
+        apply pow_le_pow_left (add_nonneg _ (zero_le_one' ℝ)) _ --two goals
+        · exact (mul_nonneg (mul_nonneg (Nat.cast_nonneg n) (norm_nonneg A)) (Nat.cast_nonneg B))
+        · rw [mul_add]
+          simp only [mul_one, add_le_add_iff_left]
+          exact Left.one_le_mul_of_le_of_le
+              (by exact_mod_cast Nat.one_le_of_lt hn) (one_le_norm_A_of_ne_zero m n A hA)
+              (Nat.cast_nonneg n)
+  _ = ((n * ‖A‖) ^ ((m / ((n : ℝ) - m)))) ^ ((n : ℝ) - m)  * (B + 1) ^ m := by
+        rw [mul_left_inj' (pow_ne_zero m (Nat.cast_add_one_ne_zero B)),
+          ← rpow_mul (mul_nonneg (Nat.cast_nonneg n) (norm_nonneg A)), ← Real.rpow_natCast]
+        congr
+        rw [div_mul_cancel₀]
+        apply sub_ne_zero_of_ne
+        exact_mod_cast Nat.ne_of_lt' hn
+  _ < (B + 1) ^ ((n : ℝ) - m) * (B + 1) ^ m := by
+        rw [mul_lt_mul_right (pow_pos (Nat.cast_add_one_pos B) m)]
+        apply rpow_lt_rpow --three goals
+        · exact rpow_nonneg (mul_nonneg (Nat.cast_nonneg n) (norm_nonneg A)) e
+        · exact Nat.lt_floor_add_one ((n * ‖A‖) ^ e)
+        · simp only [sub_pos, Nat.cast_lt]
+          exact hn
```

Comment:
```suggestion
        · simpa only [sub_pos, cast_lt] using hn
```
BTW, `Nat.cast_nonneg n` can be written as `n.cast_nonneg` etc. ("dot notation").